Test/lambdas/MatrixAssoc.dfy

```diff
@@ -0,0 +1,270 @@
+// RUN: %dafny /compile:0 /dprint:"%t.dprint" "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+type Pos = x | 0 < x witness 1
+
+/** Matrix dimension */
+const N : Pos
+
+/** Matrix indices */
+type Index = x | 0 <= x < N      // matrix indices
+
+type Matrix = Index -> Index -> int
+
+/** Function extensionality */
+lemma FunEq<X, Y>(f: X -> Y, g: X -> Y)
+    requires forall x :: f(x) == g(x)
+    ensures f == g
+{ assume false; }
+
+lemma Same(a: Matrix, b: Matrix)
+    requires forall x, y :: a(x)(y) == b(x)(y)
+    ensures a == b
+{
+    forall x ensures a(x) == b(x) {
+        FunEq(a(x), b(x));
+    }
+    FunEq(a, b);
+}
+
+/** Σ from i to n of f(i) */
+function method Sum_n(f: Index -> int, n: nat): int
+    requires n <= N
+{
+    if n == 0 then 0 else f(n - 1) + Sum_n(f, n - 1)
+}
+
+/** Σ from i to N of f(i) */
+function method Sum(f: Index -> int): int {
+    Sum_n(f, N)
+}
+
+/** Matrix multiplication */
+function method mult(a: Matrix, b: Matrix): Matrix
+{
+    ((x: Index) => (y: Index) =>
+        Sum((k: Index) => a(x)(k) * b(k)(y)))
+}
+
+/** Σ f  +  Σ g  == Σ (i => f(i) + g(i)) */
+/** proof by induction */
+lemma distr_add_n(f: Index -> int, g: Index -> int, n: nat)
+    requires n <= N
+    ensures Sum_n(f, n) + Sum_n(g, n) == Sum_n((i: Index) => f(i) + g(i), n)
+{}
+
+/** (Σ f) * x == Σ (i => f(i) * x) */
+/** proof by induction */
+lemma distr_mult_n(f: Index -> int, n: nat, x: int)
+    requires n <= N
+    ensures Sum_n(f, n) * x == Sum_n((i: Index) => f(i) * x, n)
+{}
+
+/** (Σ f) * x == Σ (i => f(i) * x) */
+lemma distr_mult(f: Index -> int, x: int)
+    ensures Sum(f) * x == Sum((i: Index) => f(i) * x)
+{
+    distr_mult_n(f, N, x);
+}
+
+/** Σ_k (Σ_l f(k,l))  ==  Σ_l (Σ_k f(k,l)) */
+/** proof by induction */
+lemma sum_assoc_n(m: Matrix, n1: nat, n2: nat)
+    requires n1 <= N && n2 <= N
+    ensures Sum_n((k: Index) => Sum_n((l: Index) => m(k)(l), n1), n2)
+            ==
+            Sum_n((l: Index) => Sum_n((k: Index) => m(k)(l), n2), n1)
+    {
+        var f  := (k: Index) => Sum_n((l: Index) => m(k)(l), n1);
+        var g  := (l: Index) => Sum_n((k: Index) => m(k)(l), n2);
+        if (n2 == 0) {
+            calc {
+                Sum_n(f, n2);
+                ==
```

Comment:
I suggest lining up the `==` with `calc`. (Something should be said in the Dafny style guide.)

Reply:
Those two test files had pretty careless formatting, I tried to reformat it into something clearer and more consistent now.